### ISSUE TITLE
persistent `the object has been modified; please apply your changes to the latest version and try again`

### ISSUE THREAD

**Issue by mflendrich** [Kong/gateway-operator#203]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Having followed the `README.md` installation instructions and applied the below yaml, I'm getting the reconciliation error from the subject repeatedly.

```yaml
kind: GatewayClass
apiVersion: gateway.networking.k8s.io/v1alpha2
metadata:
  name: kong
spec:
  controllerName: konghq.com/gateway-operator
---
kind: Gateway
apiVersion: gateway.networking.k8s.io/v1alpha2
metadata:
  name: kong
spec:
  gatewayClassName: kong
  listeners:
  - name: http
    protocol: HTTP
    port: 80
```

```console
michal@piec ~/r/K/gateway-operator (main)> kc logs -n kong-system gateway-operator-controller-manager-ddc879b87-xcrcn -f
{"level":"info","ts":1661258810.5017073,"logger":"controller-runtime.metrics","msg":"Metrics server is starting to listen","addr":"127.0.0.1:8080"}
{"level":"info","ts":1661258810.502056,"logger":"controller-runtime.webhook","msg":"Registering webhook","path":"/validate"}
{"level":"info","ts":1661258810.5020738,"logger":"setup","msg":"start webhook server","listen_address":":9443"}
{"level":"info","ts":1661258810.5020816,"logger":"setup","msg":"starting anonymous reports"}
W0823 12:46:50.502112       1 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
{"level":"info","ts":1661258810.7947516,"logger":"setup","msg":"starting telemetry manager"}
{"level":"info","ts":1661258810.794883,"logger":"setup","msg":"starting manager"}
{"level":"info","ts":1661258810.7954898,"logger":"controller-runtime.webhook.webhooks","msg":"Starting webhook server"}
{"level":"info","ts":1661258810.7955565,"msg":"Starting server","kind":"health probe","addr":"[::]:8081"}
{"level":"info","ts":1661258810.7955763,"msg":"Starting server","path":"/metrics","kind":"metrics","addr":"127.0.0.1:8080"}
I0823 12:46:50.795735       1 leaderelection.go:248] attempting to acquire leader lease kong-system/a7feedc84.konghq.com...
{"level":"info","ts":1661258810.7959898,"logger":"controller-runtime.certwatcher","msg":"Updated current TLS certificate"}
{"level":"info","ts":1661258810.79614,"logger":"controller-runtime.webhook","msg":"Serving webhook server","host":"","port":9443}
{"level":"info","ts":1661258810.796312,"logger":"controller-runtime.certwatcher","msg":"Starting certificate watcher"}
I0823 12:46:50.836472       1 leaderelection.go:258] successfully acquired lease kong-system/a7feedc84.konghq.com
{"level":"info","ts":1661258810.8366692,"msg":"Starting EventSource","controller":"dataplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"DataPlane","source":"kind source: *v1alpha1.DataPlane"}
{"level":"info","ts":1661258810.8366597,"msg":"Starting EventSource","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","source":"kind source: *v1alpha1.ControlPlane"}
{"level":"info","ts":1661258810.8367105,"msg":"Starting EventSource","controller":"dataplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"DataPlane","source":"kind source: *v1.Secret"}
{"level":"info","ts":1661258810.8367157,"msg":"Starting EventSource","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","source":"kind source: *v1.Secret"}
{"level":"info","ts":1661258810.8367257,"msg":"Starting EventSource","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","source":"kind source: *v1.ServiceAccount"}
{"level":"info","ts":1661258810.836718,"msg":"Starting EventSource","controller":"dataplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"DataPlane","source":"kind source: *v1.Service"}
{"level":"info","ts":1661258810.836734,"msg":"Starting EventSource","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","source":"kind source: *v1.Deployment"}
{"level":"info","ts":1661258810.8367403,"msg":"Starting EventSource","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","source":"kind source: *v1.ClusterRole"}
{"level":"info","ts":1661258810.8367481,"msg":"Starting EventSource","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","source":"kind source: *v1.ClusterRoleBinding"}
{"level":"info","ts":1661258810.8367398,"msg":"Starting EventSource","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","source":"kind source: *v1alpha2.Gateway"}
{"level":"info","ts":1661258810.8367653,"msg":"Starting Controller","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane"}
{"level":"info","ts":1661258810.8367698,"msg":"Starting EventSource","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","source":"kind source: *v1alpha1.DataPlane"}
{"level":"info","ts":1661258810.8367774,"msg":"Starting EventSource","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","source":"kind source: *v1alpha1.ControlPlane"}
{"level":"info","ts":1661258810.8367844,"msg":"Starting EventSource","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","source":"kind source: *v1.NetworkPolicy"}
{"level":"info","ts":1661258810.83679,"msg":"Starting EventSource","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","source":"kind source: *v1alpha1.GatewayConfiguration"}
{"level":"info","ts":1661258810.8368,"msg":"Starting EventSource","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","source":"kind source: *v1alpha2.GatewayClass"}
{"level":"info","ts":1661258810.8368063,"msg":"Starting Controller","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway"}
{"level":"info","ts":1661258810.8367536,"msg":"Starting EventSource","controller":"dataplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"DataPlane","source":"kind source: *v1.Deployment"}
{"level":"info","ts":1661258810.836915,"msg":"Starting Controller","controller":"dataplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"DataPlane"}
{"level":"info","ts":1661258810.8416328,"logger":"setup","msg":"no CA certificate Secret kong-operator-ca found, generating CA certificate"}
{"level":"info","ts":1661258810.9380016,"msg":"Starting workers","controller":"dataplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"DataPlane","worker count":1}
{"level":"info","ts":1661258810.9380639,"msg":"Starting workers","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","worker count":1}
{"level":"info","ts":1661258810.9380443,"msg":"Starting workers","controller":"controlplane","controllerGroup":"gateway-operator.konghq.com","controllerKind":"ControlPlane","worker count":1}


{"level":"error","ts":1661258856.0468998,"msg":"Reconciler error","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"bb5ede14-3c96-4ac7-9b54-2e6894c7d62e","error":"Operation cannot be fulfilled on gateways.gateway.networking.k8s.io \"kong\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1661258856.082813,"msg":"Reconciler error","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"86eb6ccb-dd53-4c8c-a8d3-3a40a7308f15","error":"Operation cannot be fulfilled on gateways.gateway.networking.k8s.io \"kong\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1661258876.9758286,"msg":"Reconciler error","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"69b3e132-9f0d-4806-9b02-3d020d14959f","error":"Operation cannot be fulfilled on gateways.gateway.networking.k8s.io \"kong\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1661258877.0275793,"msg":"Reconciler error","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"bf525157-1308-4a98-920a-428dcda2372f","error":"Operation cannot be fulfilled on gateways.gateway.networking.k8s.io \"kong\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1661258897.9746952,"msg":"Reconciler error","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"fe809bcb-c07a-4011-be00-005e389e7384","error":"Operation cannot be fulfilled on gateways.gateway.networking.k8s.io \"kong\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1661258898.0482025,"msg":"Reconciler error","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"adcd99b6-cad2-4a92-b787-bfef0867fa1f","error":"Operation cannot be fulfilled on gateways.gateway.networking.k8s.io \"kong\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1661258898.0948768,"msg":"Reconciler error","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"b64c006a-ba8e-4458-9e36-d6e1032aa064","error":"Operation cannot be fulfilled on gateways.gateway.networking.k8s.io \"kong\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"}
^C⏎    
```

### Expected Behavior

After starting KGO and deploying the yaml attached above, I expected a gateway to spin up successfully.

### Steps To Reproduce

```markdown
$ kubectl kustomize https://github.com/Kong/kubernetes-ingress-controller/config/crd | kubectl apply -f -                                                                                                                                
$ kubectl kustomize https://github.com/kubernetes-sigs/gateway-api/config/crd | kubectl apply -f -                                                                                                                                       
$ kubectl kustomize https://github.com/kong/gateway-operator/config/default | kubectl apply -f -                    

# wait for KGO pod in state "Running"

$ kubectl apply -f THE_YAML_FILE_ATTACHED_ABOVE.yaml
```

### Kong Ingress Controller version

```shell
operator efea71c1b334f4eb5d9be0e845127f54a2600834
```

### Kubernetes version

```shell
snap install microk8s --channel 1.23/stable --classic

Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.4", GitCommit:"95ee5ab382d64cfe6c28967f36b53970b8374491", GitTreeState:"clean", BuildDate:"2022-08-18T02:29:34Z", GoVersion:"go1.18.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.9-2+88a2c6a14e7008", GitCommit:"88a2c6a14e7008ae9577b448c50091f44b6be572", GitTreeState:"clean", BuildDate:"2022-07-13T18:05:30Z", GoVersion:"go1.17.10", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by mflendrich** [comment]
I'm not able to reproduce this error. Might be ready to close.

**Comment by mflendrich** [comment]
This message makes it to the non-debug logs though, therefore suggesting that something's going wrong with CP/DP provisioning where it isn't.

**Comment by mlavacca** [comment]
If the error is not harmful in any way and the Gateway gets eventually reconciled, we could silent this kind of error log simply requeuing without error if a conflict happens, we do that in many places in KGO.

**Comment by shaneutt** [comment]
> If the error is not harmful in any way and the Gateway gets eventually reconciled, we could silent this kind of error log simply requeuing without error if a conflict happens, we do that in many places in KGO. update conflicts may happen shrug

I would like to understand better why/how we can get into a state where the reconcilation is _repeatedly_ getting a 409. This may suggest we have broken logic somewhere that is "stuck" in a spec or status update loop of the object.

**Comment by shaneutt** [comment]
I just tested this locally, and I was able to re-create the log entries, but the result was

**Comment by mlavacca** [comment]
I think the problem can be here:
https://github.com/Kong/gateway-operator/blob/main/controllers/gateway_controller.go#L159-L178

The status is updated in `ensureGatewayMarkedReady`, and then it is immediately updated again in `r.updateStatus`. So, two very quickly status updates I think may produce conflicts. WDYT?

**Comment by shaneutt** [comment]
Ah yeah, we should definitely not do that. It should use the `created, err` or `updated, err` signature pattern and indicate to the caller whether a write op was done on the API, and if so it should re-queue. Good looking out, let's get a patch in for that. :+1:

**Comment by mlavacca** [comment]
Yep, I'll create a PR for this one

**Comment by mlavacca** [comment]
closed by #226
